File: src/snowflake/snowflake.ts
/**
 * js 表示的整数位 2<< 53 -1 但是生成的 id 位 64位 也就是 2<<64-1
 * js number类型精度不够,  需要使用bigint类型
 */

// Wed Jan 01 2020 00:00:00 GMT+0800 (中国标准时间)
const epoch = 1577808000000n;
// 时间戳占 41位
const time = 41n;
// 机器id占 10位
const work = 10n;
// 序列号id 12位
const sequence = 12n;
// 时间戳左移位数
const timeShift = work + sequence;
// 机器id左移位数
const workShift = sequence;
// 机器id最大值 10bit 最大值  4095
const maxWork = -1n ^ (-1n << work);
// 序列号id最大值 12bit 最大值 1023
const maxSequence = -1n ^ (-1n << sequence);

class SnowflakeWorker {
  // 时间戳
  private lastTime: bigint;
  // 序列号
  private sequenceId: bigint;
  // 固定的机器id
  private readonly workerBitValue: bigint;

  constructor(workId = 0n) {
    workId = typeof workId === 'bigint' ? workId : BigInt(workId);
    if (workId > maxWork) {
      throw new Error('Machine ID exceeds maximum limit');
    }
    this.workerBitValue = workId << workShift;
    this.lastTime = BigInt(new Date().getTime());
    this.sequenceId = 0n;
  }

  public generate() {
    let cur = BigInt(new Date().getTime());
    if (this.lastTime === cur) {
      this.sequenceId = (this.sequenceId + 1n) & maxSequence;
      if (this.sequenceId === 0n) {
        while (BigInt(new Date().getTime()) <= this.lastTime) {
          cur = BigInt(new Date().getTime());
        }
      }
    } else {
      this.sequenceId = 0n;
    }
    // 下面这两个不好测试
    if (cur < this.lastTime) {
      throw new Error('Date is error!');
    }
    this.lastTime = cur;
    return ((cur << timeShift) & epoch) | this.workerBitValue | this.sequenceId;
  }
}

export class Snowflake {
  static Worker = SnowflakeWorker;

  static getEpoch(): bigint {
    return epoch;
  }

  /**
   * 从id中解析出workerId
   * @param id
   */
  static getWorkerId(id: bigint): bigint {
    return (id & (maxWork << workShift)) >> workShift;
  }

  /**
   * 从id中解析出时间戳
   * @param id
   */
  static getTimeId(id: bigint): bigint {
    return (id >> timeShift) + epoch;
  }

  /**
   * 从id中解析出递增id
   * @param id
   */
  static getSequenceId(id: bigint): bigint {
    return id & maxSequence;
  }
}

File: go/snowflake/snowflake.go
package snowflake

import (
    "encoding/base64"
    "github.com/pkg/errors"
    "strconv"
    "time"
)

const (
    // 2019-11-07T09:58:52.370Z
    epoch = 1573120745213
    // 时间戳占 41位
    timeBits = 41
    // 机器id占 10位
    workerBits = 10
    // 序列号id 12位
    sequenceBits = 12
    // 时间戳左移位数
    timeShift = workerBits + sequenceBits
    // 机器id左移位数
    workerShift = sequenceBits
    // 机器id最大值 10bit 最大值  4095
    maxWorker = -1 ^ (-1 << workerBits)
    // 序列号id最大值 12bit 最大值 1023
    maxSequence = -1 ^ (-1 << sequenceBits)
)

type Id int64

type Worker struct {
    lastTime int64
    workerId int64
    number   int64
}

func NewWorker(workerId int64) (*Worker, error) {
    // 要先检测workerId是否在上面定义的范围内
    if workerId < 0 || workerId > maxWorker {
        return nil, errors.New("Worker ID excess of quantity")
    }
    // 生成一个新节点
    return &Worker{
        lastTime: 0,
        workerId: workerId,
        number:   0,
    }, nil
}

func (w *Worker) Generate() (Id, error) {
    now := time.Now().UnixNano() / 1e6
    if w.lastTime == now {
        w.number = (w.number + 1)&maxSequence
        if w.number == 0 {
            // 如果当前工作节点在1毫秒内生成的ID已经超过上限 需要等待1毫秒再继续生成
            for now <= w.lastTime {
                now = time.Now().UnixNano() / 1e6
            }
        }
    } else {
        w.number = 0
    }
    // 如果当前时间小于上一次ID生成的时间戳，说明系统时钟回退过这个时候应当返回错误
    if now < w.lastTime {
        return 0, errors.New("System time error")
    }
    w.lastTime = now
    ID := Id(((now - epoch) << timeShift) | (w.workerId << workerShift) | (w.number))
    return ID, nil
}

func (f Id) Time() int64 {
    return (int64(f) >> timeShift) + epoch
}

func (f Id) Worker() int64 {
    return int64(f) & (maxWorker << workerShift) >> workerShift
}
func (f Id) Sequence() int64 {
    return int64(f) & maxSequence
}
func (f Id) Bytes() []byte {
    return []byte(f.String())
}
func (f Id) Base64() string {
    return base64.StdEncoding.EncodeToString(f.Bytes())
}

func (f Id) String() string {
    return strconv.FormatInt(int64(f), 10)
}
